feat(plugin): new Row Based Editor

zewa666 opened the pull request:
this adds the new row based edit plugin

closes #1299

ghiscoding commented:
for visibility could you add the animated gif you demoed in the issue in here for anyone who opens the PR to see more details. Thanks

zewa666 commented:
yep I'll create one when writing the docs and attach it as comment. but first those nasty final two lines need coverage 😅

ghiscoding commented on packages/common/src/extensions/slickRowBasedEdit.ts:
```diff
@@ -0,0 +1,596 @@
+      this._gridService
+    ) {
+      if (this._addonOptions?.actionButtons?.deleteButtonPrompt) {
+        if (
```
is it more readable to merge these 2 `if` into a single one? ... I guess it's ok as is since I see the next few conditions are much longer, so we might as well stay consistent

ghiscoding commented on packages/common/src/extensions/slickRowBasedEdit.ts:
```diff
@@ -0,0 +1,596 @@
+              'button-style padding-1px mr-2'
+            } action-btns ` + BTN_ACTION_EDIT,
+          title:
+            options.rowBasedEditOptions?.actionButtons?.editButtonTitle ||
```
I also usually have a set of Locale translation keys for I18N, we could probably add them in the future though `editButtonTitleKey`, `deleteButtonTitleKey` and `updateButtonTitleKey`... (e.g.: `updateButtonTitleKey: 'UPDATE'`)

It's a lot of work to make everyone happy 😅 

but that might be ok to skip that since it's only for buttons?

zewa666 replied:
usually I'm the one being nitpicky about i18n ;) good catch

ghiscoding commented on packages/common/src/services/extension.service.ts:
```diff
@@ -162,6 +166,21 @@ export class ExtensionService {
+        this._rowBasedEdit = this._rowBasedEdit || new SlickRowBasedEdit(this.pubSubService, this.gridOptions.rowBasedEditOptions);
+        const gridService = this.lazyGridService?.();
+        if (!gridService) {
+          throw new Error('[Slickgrid-Universal] requires a GridService to be configured and available');
```
I think we should add RowBaseEdit somewhere in the error message to advise the throw caller to the user

ghiscoding commented on packages/common/src/services/extension.service.ts:
```diff
@@ -62,6 +65,7 @@ export class ExtensionService {
     protected readonly sortService: SortService,
     protected readonly treeDataService: TreeDataService,
     protected readonly translaterService?: TranslaterService,
+    protected readonly lazyGridService?: () => GridService
```
nice name since it really is lazy loaded, I'm still impressed by this approach 😉

zewa666 replied:
yeah lazy resolvers are a great thing over at C# standard DI and also great in Aurelia. Its a cool way to push initialization of transient tokens all the way to the point when they are actually needed.

In Angular its a bit cumbersome with Lazy Module loading but serves the same purpose.

the async nature on the other hand makes it a nice escape hatch for circular refs ;)

ghiscoding commented:
> yep I'll create one when writing the docs and attach it as comment. but first those nasty final two lines need coverage 😅

feel to use `istanbul ignore` if they're too hard to cover, I started using that for some core files which I'm not sure how to reach certain lines 😅

ghiscoding commented on packages/common/src/extensions/__tests__/slickRowBasedEdit.spec.ts:
```diff
@@ -0,0 +1,824 @@
+  });
+
+  it('should warn the user when autoEdit is not set or false and turn it on', () => {
+    const consoleSpy = jest.spyOn(console, 'warn');
```
I just saw in the GitHub Action that this console is showing up, to avoid showing them in Jest, you can (should) use the following
```ts
const consoleSpy = jest.spyOn(global.console, 'warn').mockReturnValue();
```
![image](https://github.com/ghiscoding/slickgrid-universal/assets/643976/fafe8207-22e1-496c-ac18-e41daf98e010)

zewa666 commented on examples/vite-demo-vanilla-bundle/src/examples/example22.ts:
```diff
@@ -0,0 +1,223 @@
+  }
+
+  toggleSingleMultiRowEdit() {
+    this.sgb.gridOptions = this.sgb.gridOptions = {
```
the double assignment is just a mistake. dunno how I managed to pull that off 😅

I dunno I just saw the other examples using sgb.gridOptions hence continuing in the same vein. I'll check up on that

zewa666 replied:
the approach with this.sgb.gridOptions and assigning it to this.gridOptions is the same as in example 11, 12, 14 and 19. I've just removed the wrong double assignment

zewa666 commented on examples/vite-demo-vanilla-bundle/src/examples/example22.ts:
```diff
@@ -0,0 +1,223 @@
+  return new Promise((resolve) => {
+    setTimeout(() => {
+      resolve(new Response(JSON.stringify({ status: 200, message: 'success' })));
+    }, 500);
```
its just one time to simulate the save delay, but nevertheless good idea to reduce it in case of testing

zewa666 commented on package.json:
```diff
@@ -95,5 +95,8 @@
     "url": "https://ko-fi.com/ghiscoding"
+  },
+  "prettier": {
+    "singleQuote": true
```
I will as last cleanup but I'd ask that once the PR is ready you give it a try by yourself and tweak the value to your likings

zewa666 commented on packages/common/src/extensions/__tests__/slickRowBasedEdit.spec.ts:
```diff
@@ -0,0 +1,824 @@
+    gridStub.getOptions.mockReturnValue({});
+
+    expect(() => plugin.init(gridStub, gridService)).toThrow(
+      /(enableCellNavigation = true)/
```
say thanks to Copilot I didnt notice either 😅

zewa666 commented on packages/common/src/extensions/__tests__/slickRowBasedEdit.spec.ts:
```diff
@@ -0,0 +1,824 @@
+    });
+
+    [-1, 0, 2].forEach((position) => {
+      it(
```
no actually I didnt. looks good. most of the times I'm fine with the simple loop but that gherkin style table they support looks quite convenient for a few use cases

zewa666 commented on packages/common/src/extensions/slickRowBasedEdit.ts:
```diff
@@ -0,0 +1,596 @@
+            if (existingBeforePasteCellHandler) {
+              userResult = existingBeforePasteCellHandler(e, args);
+
+              if (userResult === false) {
```
yep I wanted to explicitely make clear that this is the original user defined rules result as the handler name might not suggest that straight away. I prefer an additional variable over comments in these cases. but I can remove it if it bothers you

zewa666 commented on packages/common/src/extensions/slickRowBasedEdit.ts:
```diff
@@ -0,0 +1,596 @@
+    );
+
+    this._grid.invalidate();
+    this._grid.render();
```
yep just copied that behavior from other places. I thought you meant the double call explicitely

zewa666 commented on packages/common/src/extensions/slickRowBasedEdit.ts:
```diff
@@ -0,0 +1,596 @@
+      minWidth: 70,
+      width: 75,
+      maxWidth: 75,
+      excludeFromExport: true,
```
good catch. actually I think instead of defining width and others we should accept `actionColumnConfig: Partial<Column<any>> and spread that over the previously applied. this way users can override and extens whatever they want

zewa666 commented on packages/common/src/extensions/slickRowBasedEdit.ts:
```diff
@@ -0,0 +1,596 @@
+          return;
+        }
+
+        this._grid.invalidate();
```
oh thats great to know. looked for sonething like that already a couple of times

zewa666 commented on packages/common/src/extensions/slickRowBasedEdit.ts:
```diff
@@ -0,0 +1,596 @@
+    }
+  }
+
+  protected undoRowEdit(item: any) {
```
actually I think this might make a great extension by itself. UndoRedoPlugin sort of which than could expose hooks and events for other plugins to consume/drive features.

that way there would be one feature-rich well tested implementation which users could also take for their own apps.

ghiscoding replied:
yeah that's another good idea, future, future...

zewa666 commented:
the best thing about prettier is its so damn conservative and restrictive. the worst thing with that is that feelings get hurt. I personally grew so tired of formatting discussions that it became like a godsend tool to me.

but I'm happy to change to whatever style you prefer. I just werent able to determine clear paths for certain constructs.

so I'd like to turn this around, once the PR is done feel free to reformat the whole file and I'll turn off prettier for this file like all the others

ghiscoding commented:
@zewa666 yeah I think that I won't use Prettier in this project, it's getting quite large and I don't feel like reformatting the whole thing and revisit all test coverage. However, I do use Prettier in newer project, it's easier to maintain when the project initially started with it and when it isn't too large like the project. I'm not expecting to change the format for new files but yeah if you modify some existing file then it's probably better to disable Prettier or at the minimum enable it for "changed lines" instead of the whole file.

Just out of curiosity for my schedule and I don't want to be pushy but are you expecting to finish the PR by tomorrow? I'd like to go ahead with a new release tomorrow and I'm also guessing that a new PR that I just made might conflicts with yours so I'll wait a bit... but anyway even if it's not by tomorrow, I can still push later next week, no biggy 

Thanks a ton for this new contribution, this is for sure the biggest PR I've received since the creation of the project 🚀 ... I feel a little less lonely to maintain the project now (finally someone who takes time to read the code 😝) hehe

zewa666 commented:
no hard feelings on that. yeah just used it for the new stuff, i dont use it in the other PRs or touching existing files.

No I dont think I'll make it until tomorrow. i'd like to properly finish the docs and your feedback and get in the couple of last details like proper coloring. If you have any conflicting PR just get them in so I can merge and resolve conflicts. I'll be done next week but again no urgency in getting this released.

this plugin was meant for getting into the codebase a little bit deeper and understanding how things are connected. I feel actually super confident after what I'm seeing. the architecture is really flexible and allows for tons of extension points. plus you've been extremely helpful with tips and guidance. thanks again for that.

ghiscoding commented on packages/common/src/extensions/slickRowBasedEdit.ts:
```diff
@@ -0,0 +1,596 @@
+  }
+
+  protected onBeforeEditCellHandler = (
+    _e: Event,
```
```suggestion
    _e: SlickEventData,
```
because of recent PR #1327

ghiscoding commented:
Alright I'll go ahead and merge PR #1327 and hopefully won't cause too much trouble on your side, the PR is about using the correct `SlickEventData` when subscribing with `SlickEventHandler`, it was previously `.subscribe((e: any, args: ArgType) => {})` but now it's `(e: SlickEventData, ...)` which is the correct type, that required a lot of refactoring though which is why ~I'm expecting conflicts~ (nope it didn't add conflicts, great, but just 1 small change to do `_e: SlickEventData` on your side)


> this plugin was meant for getting into the codebase a little bit deeper and understanding how things are connected. I feel actually super confident after what I'm seeing. the architecture is really flexible and allows for tons of extension points. plus you've been extremely helpful with tips and guidance. thanks again for that.

I like to see that kind of comments :)  I do love the project, the person who created SlickGrid eventually stopped working on it and got hired by Google, his way of structuring the project was quite ahead of its time (it was pretty much the precursor of ES Module but using jQuery, which was pretty awesome at the time), it's super customizable and I'm trying to keep it that way as much as possible, I also see a lot more stars on the original `6pac/slickgrid` since we migrated to native & TypeScript. It's more work for me maintain (but it's good for visibility, more users, for example the latest PR that I've mentioned was an issue opened on that project that I've pushed in here too)

Next week is fine, I'll write some more unit tests (98% now) lol... enjoy your weekend ⛄

zewa666 commented:
I've went through your feedback and two things are open.

1. I didn't get what you meant was necessary to do for the i18n keys. Perhaps you could point me to an example?
2. The invlidateRow wasn't working as it introduced side-effects (empty row) so I kept the invalidate for now. Perhaps you want to look into that one

ghiscoding commented:
> I've went through your feedback and two things are open.
> 
> 1. I didn't get what you meant was necessary to do for the i18n keys. Perhaps you could point me to an example?
> 2. The invlidateRow wasn't working as it introduced side-effects (empty row) so I kept the invalidate for now. Perhaps you want to look into that one

Well I wasn't sure if these are used for button tooltips only? If that's the case then I guess we could skip that task entirely, I usually add locale translation keys as extra options for users who use I18N and typically it's set so that if the system finds that the property with a `Key` suffix is defined then it uses it, or if not then use default option. For example, if column picker option `columnTitleKey` is provided and `enableTranslate` is enabled then we use the key for translation, but if not we revert to using `columnTitle`

So I usually provide 2 options, the locale is with the `Key` suffix

https://github.com/ghiscoding/slickgrid-universal/blob/a9cb8ee3f1a5da4249851e5b701b027b3f72ad26/packages/common/src/interfaces/gridMenuOption.interface.ts#L10-L14

then in the code you can use `extensionUtility.translateItems` by providing an object array with the prop name (like `title`) to look for and the util will search for associated translation key by adding `Key` suffix to prop name (like `titleKey`) and translate it 

https://github.com/ghiscoding/slickgrid-universal/blob/a9cb8ee3f1a5da4249851e5b701b027b3f72ad26/packages/common/src/extensions/slickColumnPicker.ts#L174-L178

zewa666 commented:
I like the idea of the overloading translation key. I'll definitely add it. yeah its just a tooltip but nevertheless it would this way follow the same pattern other plugins use.

zewa666 commented:
aside from the two above I'd also like to ask you whether you could give the scss files a look. I'm not sure I caught all the places or if something else doesnt conform to the base idea.

ghiscoding commented on packages/common/src/styles/slick-bootstrap.scss:
```diff
@@ -107,6 +107,27 @@
+        &:hover {
+          background-color: var(--slick-row-based-edit-editmode-hover-bgcolor, $slick-row-based-edit-editmode-hover-bgcolor);
+
+          & .active {
```
I don't think the space is necessary, right? `&.active` if it is necessary then remove the `&` I think

ghiscoding commented on packages/common/src/styles/slick-grid.scss:
```diff
@@ -84,6 +87,26 @@
+      &:hover {
+        background-color: var(--slick-row-based-edit-editmode-hover-bgcolor, $slick-row-based-edit-editmode-hover-bgcolor);
+
+        & .active {
```
again is the space necessary?

zewa666 commented:
ok I'm still not sure I really understand how to approach the i18n part.
Here's what I've tried so far:
[i18n.patch](https://github.com/ghiscoding/slickgrid-universal/files/13929602/i18n.patch)

the mentioned function `translateItems` isn't really suitable (array based) plus it doesn't really do anything besides calling translate.

ghiscoding commented:
ah sorry that wasn't a good explanation, without looking at the patch you linked (I'm supposed to be off on Sundays ;)). So anyway, I'll assume you might be looking for code when the locale changes, below is the code that calls that. 

https://github.com/ghiscoding/slickgrid-universal/blob/1d759f772fa3f1334f66b1cb76c37a1134f5c44d/packages/vanilla-bundle/src/components/slick-vanilla-grid-bundle.ts#L733-L752

So you probably need to add a method that will be called to from the callback above. 

If that is still not helping, just let me know and I'll take a proper look at the patch

zewa666 commented:
ok think I got it working. There was still a nasty bug with Cypress but I got a workaround to force drop the old cached items.
Next up ... writing docs

ghiscoding commented on examples/vite-demo-vanilla-bundle/public/i18n/fr.json:
```diff
@@ -100,5 +100,7 @@
-  "X_DAY_PLURAL": "{{x}} journée{{plural}}"
+  "X_DAY_PLURAL": "{{x}} journée{{plural}}",
+  "RBE_BTN_UPDATE": "Mettre à jour la ligne actuelle",
+  "RBE_BTN_CANCEL": "Annuler la ligne actuelle"
```
this is just a French thing but I think it would be better with `ligne courante` because `ligne actuelle` actually means `actual line` which is not exactly the best to use. I know it's Google Translate result that you used but I think it's wrong hehe. BTW, there's 2 of them to change 


![image](https://github.com/ghiscoding/slickgrid-universal/assets/643976/bd9315f4-53d0-4f89-b728-1024c491719d)

ghiscoding commented on packages/common/src/styles/slick-bootstrap.scss:
```diff
@@ -107,6 +107,27 @@
+          }
+        }
+
+        & .active {
```
I think you missed that one when I mentioned it couple days ago

ghiscoding commented on packages/common/src/styles/slick-grid.scss:
```diff
@@ -84,6 +87,26 @@
+        }
+      }
+
+      & .active {
```
same as previous, `&` can be removed

ghiscoding commented on packages/common/src/extensions/slickRowBasedEdit.ts:
```diff
@@ -229,6 +239,45 @@ export class SlickRowBasedEdit {
+          this.gridOptions.rowBasedEditOptions?.actionButtons?.updateButtonTitleKey
+        )) ||
+      this.gridOptions.rowBasedEditOptions?.actionButtons?.updateButtonTitle ||
+      'Update the Row';
```
I see a lot of the same code repeated, perhaps we could migrate this to an external protected function? Something along these lines

```ts
protected getTitleOrDefault(key: string, defaultTitle: string) {
  const actionBtnOptions = this.gridOptions.rowBasedEditOptions?.actionButtons; 
  const hasTranslateService = this.extensionUtility.translaterService?.translate;

  // another approach might be to move the line above to GETTERs in the class?
  
  return hasTranslateService ? actionBtnOptions[key] : defaultTitle;
}

this.btnUpdateTitle = this.getTitleOrDefault('updateButtonTitleKey', 'Update the Row');
```

I should probably do that in some other extensions like the GridMenu that I think I did something like you did and I think that this could make the code more readable.

ghiscoding commented:
this is out of context but, it's a nice story... so I keep adding unit tests for the core files and it's not exactly a fun exercise but I keep find and fixing small bugs and remove a bunch of unused or unreachable code. So it does seem to be worth the work and a worth exercise (now at 99.4% coverage, the end of the tunnel is in sight woohoo)

zewa666 commented:
yeah its great when you see that type of payoff. I wasnt able to invest time the last few days sadly but hope to manage it in the next few ones.

zewa666 commented:
hey with the latest commit I've added again a bit of TS type inference helpers.

![image](https://github.com/ghiscoding/slickgrid-universal/assets/2861414/99959185-dd65-4733-92c9-b7ed02f470b6)

Way back than I've added already some to create the Column defintion type. So my question is whether it's ok to leave them all sprinkled around or you'd want to co-locate them all in one ts-utils.ts file sort of

ghiscoding commented:
> hey with the latest commit I've added again a bit of TS type inference helpers.
> 
> ![image](https://private-user-images.githubusercontent.com/2861414/297505444-99959185-dd65-4733-92c9-b7ed02f470b6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU1MjM0NzcsIm5iZiI6MTcwNTUyMzE3NywicGF0aCI6Ii8yODYxNDE0LzI5NzUwNTQ0NC05OTk1OTE4NS1kZDY1LTQ3MzMtOTJjOS1iN2VkMDJmNDcwYjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDExNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMTdUMjAyNjE3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDg3ZWRjMmNkNDYzOTZkMjlhMTYwNzZiZGRlODU1MDg1YWU5YmFmYjY3YmZiMGZjYjkyMTUxZjI5ZmExMDUxOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.0iKsNePGCn2TYf-6icKN9w-ckYmTl_qtpCz2oEgn6Uc)
> 
> Way back than I've added already some to create the Column defintion type. So my question is whether it's ok to leave them all sprinkled around or you'd want to co-locate them all in one ts-utils.ts file sort of

ah yeah I saw the inference type you added, at first glance I thought it was overkill since it's only used once hahah but nonetheless it's quite powerful and you can't go wrong with that approach. Feel free to put it where you think is best, when I started the project, I used to create a file per interface but I see my folder is huge now and I started grouping them by feature lately depending on what it is. So you could add it to `interfaces/rowBasedEditOption.interface.ts` if it's only used for that or create something like `inference.helper` or whatever and if that is what you go with then maybe move the ones you've done in the past from `column.interface`. I don't have any preferences for that, go with what you think is best

https://github.com/ghiscoding/slickgrid-universal/blob/03cad4a34bf13a8e1342306f9210525f5025321f/packages/common/src/interfaces/column.interface.ts#L18-L31

zewa666 commented:
yeah I'm a bit nervous when magic strings are floating around as my future self isn't that mindful of knowing what the intention was. So it serves as a real well gatekeeper to focus the type and be more expressive (all of that just to save one line of comment 🤣 )

Nevertheless the docs are now also added. A quick review on those would be very appreciated.

ghiscoding commented on docs/TOC.md:
```diff
@@ -54,6 +54,7 @@
 * [Pinning (frozen) of Columns/Rows](grid-functionalities/frozen-columns-rows.md)
 * [Row Selection](grid-functionalities/Row-Selection.md)
 * [Tree Data Grid](grid-functionalities/Tree-Data-Grid.md)
+* [Row Based Editing Plugin](grid-functionalities/Row-based-edit.md)
```
doesn't need to be changed but just as a side note, I want to use only lowercase for new docs. I haven't changed them all because I'm on Windows and Git doesn't detect case change by default (there's some extra git command to run which I forgot), so just a reminder for future docs...

ghiscoding commented on docs/grid-functionalities/Row-based-edit.md:
```diff
@@ -0,0 +1,70 @@
+  - [Adding a Column dynamically is removing the Row Selection, why is that?](#adding-a-column-dynamically-is-removing-the-row-selection-why-is-that)
+
+### Description
+The Row based editing plugin makes it possible to keep the grid readonly except for rows which the user explicitely toggles into editmode.
```
is "editmode" a property? If so, wrap it in backtick or if not then maybe separate it in 2 words?

ghiscoding commented on docs/grid-functionalities/Row-based-edit.md:
```diff
@@ -0,0 +1,70 @@
+If the [Excel Copy Buffer Plugin](grid-functionalities/excel-copy-buffer.md) is configured, the Row based edting pluging will override it's behavior by denying pastes on all cells not within a editmode row. Nevertheless, any existing `BeforePasteCellHandler` will be respected.
+
+## How the plugin works
+The idea of the plugin is to focus the users editing experience on specific individual rows and and save them individually. This is achieved by letting the user toggle one or more rows into editmode. Now changes can be made to those rows and will be highlighted and tracked. The user may cancel the editmode at any time and revert all cells changes. If the save button is pressed on the other hand a ...
```
"...and toggles back into readonly mode."  so just to confirm, when this edit mode is used, the grid itself cannot be edited the typical way (click on any cell), we really need to use the edit mode, is that it? We might need to add a sentence explaining what I just asked, it wasn't clear to me

zewa666 replied:
exactly. i've added that in the description at the begin. do you want to repeat the info here to be on the safe side?

ghiscoding replied:
yeah might as well to be clearer since I didn't notice the top description

ghiscoding commented on docs/column-functionalities/Editors.md:
````diff
@@ -369,4 +369,8 @@ For SalesForce it's nearly the same, the only difference is that we add our even
+```
+
+## Turning individual rows into editmode
+Using the [Row Based Editing Plugin](../grid-functionalities/Row-based-edit.md) you can let the user toggle either one or multiple rows into editmode, keep track of cell changes and either discard or save them on an individual basis using a custom onBeforeSave hook.
````
probably better to escape `onBeforeSave` hook with backtick

zewa666 commented:
alright I think I got everything in that I've planned for. How do you want to proceed from here?

ghiscoding commented:
looking good.... let's hit the Merge button 🚀 

I think the next steps will be to replicate the new Example into the external repos (Angular-Slickgrid, ...), did you wanted to me to do that or did you wanted to do it too (for Angular I mean)?

zewa666 commented:
I can do so for Angular whenever you release slickgrid universal and bump it as dependency in Angular-slickgrid.
